src/client/metric.rs

```diff
@@ -0,0 +1,528 @@
+use byteorder::WriteBytesExt;
+use memmap::{Mmap, MmapViewSync, Protection};
+use std::ffi::CString;
+use std::io;
+use std::io::{Cursor, Write};
+use std::mem;
+
+const ITEM_BIT_LEN: usize = 10;
+
+pub const I32_METRIC_TYPE_CODE: u32 = 0;
+pub const U32_METRIC_TYPE_CODE: u32 = 1;
+pub const I64_METRIC_TYPE_CODE: u32 = 2;
+pub const U64_METRIC_TYPE_CODE: u32 = 3;
+pub const F32_METRIC_TYPE_CODE: u32 = 4;
+pub const F64_METRIC_TYPE_CODE: u32 = 5;
+pub const STRING_METRIC_TYPE_CODE: u32 = 6;
+
+/// Generic type for any Metric's value
+pub trait MetricType {
+    /// Returns the MMV metric type code
+    fn type_code(&self) -> u32;
+    /// Writes the byte representation of the value to a writer.
+    ///
+    /// For integer and float types, the byte sequence is little endian.
+    ///
+    /// For the string type, the UTF-8 byte sequence is suffixed with a null byte.
+    fn write_to_writer<W: WriteBytesExt>(&self, writer: &mut W)
+        -> io::Result<()>;
+}
+
+macro_rules! impl_metric_type_for (
+    ($typ:tt, $base_typ:tt, $type_code:expr) => (
+        impl MetricType for $typ {
+            
+            fn type_code(&self) -> u32 {
+                $type_code
+            }
+
+            fn write_to_writer<W: WriteBytesExt>(&self, mut w: &mut W)
+            -> io::Result<()> {
+                w.write_u64::<super::Endian>(
+                    unsafe {
+                        mem::transmute::<$typ, $base_typ>(*self) as u64
+                    }
+                )
+            }
+
+        }
+    )
+);
+
+impl_metric_type_for!(i32, u32, I32_METRIC_TYPE_CODE);
+impl_metric_type_for!(u32, u32, U32_METRIC_TYPE_CODE);
+impl_metric_type_for!(i64, u64, I64_METRIC_TYPE_CODE);
+impl_metric_type_for!(u64, u64, U64_METRIC_TYPE_CODE);
+impl_metric_type_for!(f32, u32, F32_METRIC_TYPE_CODE);
+impl_metric_type_for!(f64, u64, F64_METRIC_TYPE_CODE);
+
+impl MetricType for String {
+    fn type_code(&self) -> u32 {
+        STRING_METRIC_TYPE_CODE
+    }
+
+    fn write_to_writer<W: Write>(&self, mut writer: &mut W) -> io::Result<()> {
+        writer.write_all(CString::new(self.as_str())?.to_bytes_with_nul())
+    }
+}
+
+#[derive(Copy, Clone)]
+/// Scale for the space component of a unit
+pub enum SpaceScale {
+    /// byte
+    Byte = 0,
+    /// kilobyte (1024 bytes) 
+    KByte,
+    /// megabyte (1024^2 bytes)
+    MByte,
+    /// gigabyte (1024^3 bytes)
+    GByte,
+    /// terabyte (1024^4 bytes)
+    TByte,
+    /// petabyte (1024^5 bytes)
+    PByte,
+    /// exabyte (1024^6 bytes)
+    EByte
+}
+
+#[derive(Copy, Clone)]
+/// Scale for the time component of a unit
+pub enum TimeScale {
+    /// nanosecond
+    NSec = 0,
+    /// microsecond
+    USec,
+    /// millisecond
+    MSec,
+    /// second
+    Sec,
+    /// minute
+    Min,
+    /// hour
+    Hour
+}
+
+#[derive(Copy, Clone)]
+/// Scale for the count component of a unit
+pub enum CountScale {
+    One = 0
+}
+
+#[derive(Copy, Clone)]
+/// Unit for a Metric
+pub struct Unit {
```

Comment:
Unit still looks bloated with the main type being a struct and having 6 fields when only 2 are important at a time. I was thinking of something like

```rs
pub trait Unit {
	 fn pmapi_repr(&self) -> u32;
}

#[derive(Debug, Copy, Clone)]
pub enum SpaceUnit {
	Byte = 1 << 28,
	KiloByte = 1 << 28 | 1 << 16,
	MegaByte = 1 << 28 | 2 << 16,
	// ...
}

impl Unit for SpaceUnit {
	fn pmapi_repr(&self) -> u32 { *self as u32 }
}

#[derive(Debug, Copy, Clone)]
pub enum TimeUnit {
	NanoSecond = 1 << 24,
	MicroSecond = 1 << 24 | 1 << 12,
	MilliSecond = 1 << 24 | 2 << 12,
	// ...
}

impl Unit for TimeUnit {
	fn pmapi_repr(&self) -> u32 { *self as u32 }
}

#[derive(Debug, Copy, Clone)]
pub enum CountUnit {
	One = 1 << 20,
	// ...
}

impl Unit for CountUnit {
	fn pmapi_repr(&self) -> u32 { *self as u32 }
}
```